runtime/near-vm-logic/src/alt_bn128.rs

```diff
@@ -1,348 +1,169 @@
-use bn::arith::U256;
-use bn::{pairing_batch, AffineG1, AffineG2, Fq, Fq2, Fr, Group, GroupError, Gt, G1, G2};
-use borsh::{BorshDeserialize, BorshSerialize};
-use near_vm_errors::VMLogicError;
-use std::io::{self, Error, ErrorKind, Write};
+use crate::array_utils::{join_array, split_array, ArrayChunks};
+use bn::Group;
+use near_vm_errors::{HostError, VMLogicError};
 
-use crate::HostError;
+const BOOL_SIZE: usize = 1;
+const U128_SIZE: usize = 128 / 8;
+const SCALAR_SIZE: usize = U128_SIZE * 2;
+const POINT_SIZE: usize = SCALAR_SIZE * 2;
 
-const POINT_IS_NOT_ON_THE_CURVE: &str = "point is not on the curve";
-const POINT_IS_NOT_IN_THE_SUBGROUP: &str = "point is not in the subgroup";
-const NOT_IN_FIELD: &str = "integer is not less than modulus";
-
-#[inline]
-pub fn ilog2(n: u64) -> u64 {
-    assert!(n > 0);
-    63 - n.leading_zeros() as u64
+pub(crate) struct InvalidInput {
+    pub(crate) msg: String,
 }
 
-pub fn alt_bn128_g1_multiexp_sublinear_complexity_estimate(
-    n_bytes: u64,
-    discount: u64,
-) -> Result<u64, HostError> {
-    // details of computation alt_bn128 parameters are available at https://gist.github.com/snjax/90e8b3e7f5c16a983a5f6347d1d28bde
-    const A: u64 = 85158;
-    const B: u64 = 15119;
-    const C: u64 = 682573;
-    const MULTIEXP_ITEM_SIZE: u64 = std::mem::size_of::<(G1, Fr)>() as u64;
-
-    // A+C*n/(log2(n)+4) - B*n - discount
-    let n = (n_bytes)
-        .checked_add(MULTIEXP_ITEM_SIZE)
-        .ok_or(HostError::IntegerOverflow)?
-        .checked_add(MULTIEXP_ITEM_SIZE)
-        .ok_or(HostError::IntegerOverflow)?
-        / MULTIEXP_ITEM_SIZE;
+impl InvalidInput {
+    fn new(msg: &str, bad_value: &[u8]) -> InvalidInput {
+        let msg = format!("{msg}: {bad_value:X?}");
+        InvalidInput { msg }
+    }
+}
 
-    let res = A;
-    if n != 0 {
-        let growth_factor = std::cmp::min(ilog2(n), 15);
-        res.checked_add(
-            n.checked_mul(growth_factor + 3)
-                .ok_or(HostError::IntegerOverflow)?
-                .checked_add(1 << (1 + growth_factor))
-                .ok_or(HostError::IntegerOverflow)?
-                .checked_mul(C)
-                .ok_or(HostError::IntegerOverflow)?
-                / ((growth_factor + 4) * (growth_factor + 5)),
-        )
-        .ok_or(HostError::IntegerOverflow)?;
+impl From<InvalidInput> for VMLogicError {
+    fn from(err: InvalidInput) -> Self {
+        HostError::AltBn128InvalidInput { msg: err.msg }.into()
     }
+}
 
-    Ok(res.saturating_sub(
-        B.checked_mul(n)
-            .ok_or(HostError::IntegerOverflow)?
-            .checked_add(discount)
-            .ok_or(HostError::IntegerOverflow)?,
-    ))
+pub(crate) fn split_elements<const ELEMENT_SIZE: usize>(
+    data: &[u8],
+) -> Result<ArrayChunks<'_, ELEMENT_SIZE>, InvalidInput> {
+    ArrayChunks::new(data)
+        .map_err(|leftover| InvalidInput { msg: format!("{} leftover bytes", leftover.len()) })
```

Comment:
This seems like a pretty difficult to track down error message, especially knowing that we never really introduce any context later on. So figuring out how things got to here seem like would be pretty hard to uncover.

Reply:
We have a bit of context here:

https://github.com/near/nearcore/blob/603624657593ef9c70338f3cd839c46b0b525ab2/runtime/near-vm-errors/src/lib.rs#L423

But yeah, it's better if the error message mentions input length and element size, fixed. 

On a more general note:

I considered making fine grained well-typed errors here, but, as this is the stuff we store in DB and changing schemas is a pain, I decided that flexibility we gain from erasing everything into a string is probably worth it.